README.md
# uplifteval
[![](https://www.r-pkg.org/badges/version/uplifteval?color=orange)](https://cran.r-project.org/package=uplifteval)
[![](http://cranlogs.r-pkg.org/badges/grand-total/uplifteval?color=green)](https://cran.r-project.org/package=uplifteval)

## Description
`uplifteval` provides a variety of plots and metrics to evaluate the predictions of uplift models for both binary and continuous outcomes.  Naming conventions are somewhat inconsistent due to the mix of packages that have been ported into this package.

Included with the package are:

| Function      | Description   |
|:------------- |:------------- |
| plot_uplift_guelman  | A copy of the R `uplift` package's Qini chart and metrics. Binary outcomes only. |
| pl_plot | A port of the python pylift package's Qini chart and metrics. Binary outcomes only. |
| plot_uplift | A display of absolute cumulative differential response for bootstrap samples of balanced or unbalanced treatment and control groups, ordered by descending score.  Also displays a histogram of treatment and control groups to confirm if their distributions are balanced and identical.  Both binary and continuous outcomes.|


## Examples
Below is an example using the grf package to estimate uplift via a causal forest.  We then plot the results using three methods: the uplift package, the port of the pylift plot, and an alternative uplift plot. 
```
set.seed(123)

rl <- function(x){
  round(1/(1+exp(-x)))
}
n = 2000; p = 10
X = matrix(rnorm(n*p), n, p)
W = rbinom(n, 1, 0.2)
Y = rl(rl(X[,1]) * W - rl(X[,3]) * W + rnorm(n))
tau.forest = causal_forest(X, Y, W)
tau.hat = predict(tau.forest, X)

# the `uplift` plot
plot_uplift_guelman(tau.hat$predictions, W, Y)

# the port of pylift plot
plue = plUpliftEval(W.test, Y.test, tau.hat$predictions)
pl_plot(plue,
        show_practical_max = TRUE,
        show_theoretical_max = TRUE,
        show_no_dogs = TRUE,
        n_bins=20)

# the alternative uplift plot
plot_uplift(tau.hat$predictions, W, Y)

```

## Contributions
Feedback and contributions are welcome.  Please submit an [issue](https://github.com/ras44/uplifteval/issues).

## To-do
- when comparing two models, score ranges might be different, consider offering depth option
- add cost of targeting option


NEWS.md
# uplifteval 0.1.0.9000

In development.

# uplifteval 0.1.0

First release.

Available plotting functions:
- plot_uplift_guelman
- pl_plot
- plot_uplift


tests/testthat.R
library(testthat)
library(uplifteval)

test_check("uplifteval")


tests/testthat/test_pylift.R
context("placeholder tests")
library(whisker)
library(testthat)

test_that("plUpliftEval factor treatments error", {
  n <- 100;
  W <- rbinom(n, 1, 0.5)
  W <- as.factor(W)
  Y <- rbinom(n, 1, 0.5)
  p <- rnorm(n)
  expect_error(plUpliftEval(W,Y,p))
})

test_that("plUpliftEval factor outcomes error", {
  n <- 100;
  W <- rbinom(n, 1, 0.5)
  Y <- rbinom(n, 1, 0.5)
  Y <- as.factor(Y)
  p <- rnorm(n)
  expect_error(plUpliftEval(W,Y,p))
})


test_that("plUpliftEval factor predictions error", {
  n <- 100;
  W <- rbinom(n, 1, 0.5)
  Y <- rbinom(n, 1, 0.5)
  p <- as.factor(rnorm(n))
  expect_error(plUpliftEval(W,Y,p))
})

test_that("plUpliftEval different length treatments vs. predictions error", {
  n <- 100;
  W <- rbinom(n, 1, 0.5)
  Y <- rbinom(n, 1, 0.5)
  p <- as.factor(rnorm(2*n))
  expect_error(plUpliftEval(W,Y,p))
})

test_that("plUpliftEval different length treatments vs. outcomes error", {
  n <- 100;
  W <- rbinom(n, 1, 0.5)
  Y <- rbinom(2*n, 1, 0.5)
  p <- as.factor(rnorm(n))
  expect_error(plUpliftEval(W,Y,p))
})



tests/testthat/test_placeholder.R
context("placeholder tests")
library(whisker)

test_that(" 1==1 ", {

  expect_true(1==1)

})


R/plot_uplift.R
#' Creates an uplift plot of cumulative differential treatment/control outcomes
#' versus model score.  Also provides a selection of metrics: max uplift as pct
#' of total control outcome, optimum users targeted and optimum score targeting
#' range.
#'
#' @param p1 numeric vector of uplift predictions; can also be predicted outcomes
#'   for treated case (in this case p0 should contain predicted outcomes for
#'   the control case)
#' @param W binary vector {1,0} of treatment assignments
#' @param Y numeric vector of responses
#' @param ns integer number of samples per bootstrap iteration; default min(table(W))
#' @param n_bs integer number of bootstrap iterations
#' @param W_label optional labels for the treatment options (default W)
#' @param p0 optional numeric vector of predicted outcomes for control case
#' @param balanced optional boolean whether to sample equal proportions from
#'   treatment and control cases; default TRUE
#' @param replace optional boolean whether to use replacement when sampling;
#'   default TRUE
#' @param x_interval optional numeric the interval with which to split the
#' @param ... additional arguments (unused)
#'   x-axis
#'
#' @import ggplot2 graphics
#' @importFrom dplyr %>%
#'
#' @examples
#'
#' set.seed(0)
#' rl <- function(x){
#'   round(1/(1+exp(-x)))
#' }
#' n <- 2000; p <- 3
#' beta <- -0.5
#' X <- matrix(rnorm(n*p), n, p)
#' W <- rbinom(n, 1, 0.5)
#' Y <- rl(pmax(beta+X[,1], 0) * W + X[,2])
#' p1 <- 1/(1+exp(-(beta+X[,1])))
#' plot_uplift(p1, W, Y, n_bs=20, x_interval = 0.05, balanced = TRUE)
#'
#'
#' set.seed(0)
#' n <- 2000; p <- 3
#' beta <- -0.5
#' X <- matrix(rnorm(n*p), n, p)
#' W <- rbinom(n, 1, 0.5)
#' Y <- pmax(beta+X[,1], 0) * W + X[,2]
#' p1 <- 1/(1+exp(-(beta+X[,1])))
#' plot_uplift(p1, W, Y, n_bs=20, x_interval = 0.05, balanced = TRUE)
#'
#'
#' \donttest{
#' library(grf)
#' set.seed(123)
#'
#' rl <- function(x){
#'   round(1/(1+exp(-x)))
#' }
#' n = 2000; p = 10
#' X = matrix(rnorm(n*p), n, p)
#' W = rbinom(n, 1, 0.2)
#' Y = rl(rl(X[,1]) * W - rl(X[,3]) * W + rnorm(n))
#' tau.forest = causal_forest(X, Y, W)
#' tau.hat = predict(tau.forest, X)
#' plot_uplift(tau.hat$predictions, W, Y, n_bs=20, x_interval = 0.05, balanced = FALSE)
#' plot_uplift(tau.hat$predictions, W, Y, n_bs=20, x_interval = 0.05, balanced = TRUE)
#'
#' }
#' @export

plot_uplift <- function(p1,
                        W,
                        Y,
                        ns=min(table(W)),
                        n_bs = 1,
                        W_label=W,
                        p0=rep(0,length(p1)),
                        balanced = TRUE,
                        replace = TRUE,
                        x_interval = 0.1,
                        ...
                        ){

  q_pred_t <- c()
  q_pred_c <- c()
  q_resp <- c()
  q_treat <- c()
  q_treat_label <- c()
  q_iter <- c()

  if(balanced){
    for(i in c(1:n_bs)){
      inds_t0 <- sample(seq_along(W)[W==0],ns, replace = replace)
      inds_t1 <- sample(seq_along(W)[W==1],ns, replace = replace)
      q_pred_t <- c(q_pred_t, c(p1[inds_t0],p1[inds_t1]))
      q_pred_c <- c(q_pred_c, c(p0[inds_t0],p0[inds_t1]))
      q_resp <- c(q_resp, c(Y[inds_t0],Y[inds_t1]))
      q_treat <- c(q_treat, c(W[inds_t0],W[inds_t1]))
      q_treat_label <- c(q_treat_label, c(W_label[inds_t0],W_label[inds_t1]))
      q_iter <- c(q_iter, c(rep(i,ns),rep(i,ns)))
    }
  }
  else {
    for(i in c(1:n_bs)){
      print(paste0("boostrap iter: ", i))
      # subsample must be defined for a particular model
      # it must append to the above q_ variables in the calling environment via <<-
      inds <- sample(length(p1), ns, replace = replace)
      q_pred_t <- c(q_pred_t, p1[inds])
      q_pred_c <- c(q_pred_c, p0[inds])
      q_resp <- c(q_resp, Y[inds])
      q_treat <- c(q_treat, W[inds])
      q_treat_label <- c(q_treat_label, W_label[inds])
      q_iter <- c(q_iter, rep(i,ns))
    }
  }


  dif.pred <- q_pred_t - q_pred_c

  if(all(c(0,1) %in% unique(q_treat))){
    q_treat <- 2*q_treat-1
  }

  mm <- cbind(dif.pred = dif.pred, y = q_resp, ct = q_treat, ctl = q_treat_label, dif.pred_r = rank(-dif.pred), i = q_iter)

  stopifnot(all(c(-1,1) %in% unique(q_treat)))

  # Cumulatively sum the outcome by treatment and control groups
  mmo <- mm[order(-dif.pred)[],]
  mmo <- cbind(mmo, cdr = cumsum(mmo[,'y']*mmo[,'ct']))
  # show cumsum(dr) for individual bootstrap iterations
  mmo_df <- as.data.frame(mmo)

  mmo_df <- mmo_df %>%
    dplyr::group_by(i) %>%
    dplyr::arrange(-dif.pred) %>% dplyr::mutate(cdri = cumsum(y*ct))

  # Plot mean with min/max error bars, quantizing scoring (q = 10^x)
  q <- 1/x_interval
  mmo_dfs <- mmo_df %>% dplyr::group_by(dif.pred = floor(dif.pred*q)/q) %>%
    dplyr::summarize(min_cdri = min(cdri),
                     sd_cdri = sd(cdri),
                     max_cdri = max(cdri),
                     mean_cdri = mean(cdri))
  mmo_dfs <- cbind(mmo_dfs, group=rev(seq_along(mmo_dfs$mean_cdri))) # note: group is reversed
  mmo_dfs$sd_cdri[is.na(mmo_dfs$sd_cdri)] <- 0

  max_mean_uplift <- max(mmo_dfs$mean_cdri)
  max_mean_uplift

  max_group <- min(mmo_dfs$group[mmo_dfs$mean_cdri==max_mean_uplift])
  max_group
  dr_opt_min <- min(mmo_dfs$dif.pred[mmo_dfs$group<=max_group])
  dr_opt_min
  dr_opt_max <- max(mmo_dfs$dif.pred[mmo_dfs$group<=max_group])
  dr_opt_max

  mean_total_response_tc <- sum(mmo_df$y[mmo_df$ct==-1])/n_bs
  mean_total_response_tc

  diff_uplift_pct <- 100 * max_mean_uplift / mean_total_response_tc
  diff_uplift_pct

  mean_opt_users_targeted <- length(mmo_df$y[mmo_df$dif.pred>=dr_opt_min & mmo_df$ct==-1])/n_bs
  mean_opt_users_targeted
  mean_total_sampled_users <- length(mmo_df$y[mmo_df$ct==-1])/n_bs
  mean_total_sampled_users

  if(max_mean_uplift == 0){
    dr_opt_min <- dr_opt_max
    opt_users_targeted <- 0
  }

  config <- list(
    ct=0,
    tt=1,
    model_type="cf"
  )

  wr <- function(s,c){
    whisker::whisker.render(s,c)
  }

  s <- sign(mean_total_response_tc)

  p1 <- ggplot(mmo_dfs, aes(x = dif.pred, y = mean_cdri/mean_total_response_tc*s)) +
    geom_point(size = 2) +
    #geom_errorbar(aes(ymin = min_cdri, ymax = max_cdri)) +
    geom_errorbar(aes(ymin = (mean_cdri-sd_cdri)/mean_total_response_tc*s, ymax = (mean_cdri+sd_cdri)/mean_total_response_tc*s)) +
    scale_x_reverse() +
    xlab("score") +
    ylab("CDR") +
    labs(
      title=wr("Cumulative Differential Response T{{{tt}}}-T{{{ct}}}\n  model_type: {{{model_type}}}; {{{n_bs}}} subsample iterations of size {{{mean_total_sampled_users}}}\n  max mean uplift: {{{diff_uplift_pct}}}% of T{{{ct}}} result\n  optimum: {{{opt_pct_users_targeted}}}% of subjects scoring: [{{{dr_opt_min}}},{{{dr_opt_max}}}]; uplift/subject: {{{uplift_per_subject}}}", list(tt=config$tt, ct=config$ct, model_type=config$model_type, n_bs=n_bs, max_mean_uplift=round(max_mean_uplift,2), diff_uplift_pct=round(diff_uplift_pct,2), mean_opt_users_targeted=round(mean_opt_users_targeted,0), mean_total_sampled_users = mean_total_sampled_users, opt_pct_users_targeted = round(100*mean_opt_users_targeted/mean_total_sampled_users,3), dr_opt_min=round(dr_opt_min,3), dr_opt_max=round(dr_opt_max,3), uplift_per_subject = round(max_mean_uplift/mean_opt_users_targeted,3)))
    ) +
    theme(plot.title = element_text(size=8))

  p2 <- ggplot(data.frame(score=mmo_df$dif.pred, treatment=as.factor(mmo_df$ctl)), aes(x=score, fill=treatment)) +
    geom_histogram(position="dodge", bins = 100, linetype="blank") +
    scale_x_reverse() +
    theme(legend.position = c(0.9, 0.8))

  gridExtra::grid.arrange(p1, p2, nrow=2)

}


R/plUpliftEval.R
EPS = 1e-20
PLOT_TYPES = c('qini', 'aqini', 'cgains', 'cuplift', 'uplift', 'balance')


#' A non-S3 "constructor" function that returns a list representing a pylift uplift
#' eval object, PlUpliftEval.  This object contains metrics and can be used to generate plots.
#'
#' @param treatment numeric vector of treatment identifiers
#' @param outcome numeric vector of outcomes
#' @param prediction numeric vector of uplift predictions
#' @param p optional "infer", numeric, numeric vector representing treatment
#'   propensities
#' @param n_bins integer number of bins on x-axis; default 20
#' @return a list representing a pylift uplift eval object
#'
#' @export
new_PlUpliftEval <- function(treatment = integer(),
                             outcome = integer(),
                             prediction = numeric(),
                             p = "infer",
                             n_bins = 20){
  # Counts.
  counts = get_counts(treatment, outcome, p)
  tc_counts = get_tc_counts(counts$Nt1o1, counts$Nt0o1, counts$Nt1o0, counts$Nt0o0)
  counts = merge(counts, tc_counts)

  # Calculate maximal curves.
  count_functions = c('get_overfit_counts', 'get_no_sure_thing_counts', 'get_no_sleeping_dog_counts')
  count_names = c('max', 'pmax', 'nosdmax')
  curve_functions = c('maximal_qini_curve', 'maximal_uplift_curve', 'maximal_cuplift_curve')
  curve_names = c('qini', 'uplift', 'cuplift')

  curve_count_xys <- list()

  for (ico in seq(1:length(count_functions))){
    for(icu in seq(1:length(curve_functions))){
      cuf <- match.fun(curve_functions[icu])
      cof <- match.fun(count_functions[ico])
      xy <- cuf(cof, counts$Nt1o1, counts$Nt0o1, counts$Nt1o0, counts$Nt0o0)
      curve_count_xys[[paste0(curve_names[icu],'_',count_names[ico])]] <- xy
    }
  }

  # Calculate Q, q1, q2 scores and other metrics.
  scores = get_scores(treatment, outcome, prediction, p)

  list(
    treatment = treatment,
    outcome = outcome,
    prediction = prediction,
    counts = counts,
    p = p,
    n_bins = n_bins,
    curve_count_xys = curve_count_xys,
    scores = scores
  )
}


#' A helper for the new_PlUpliftEval function that validates the treatment,
#' outcome, prediction, p, and n_bins arguments.
#'
#' @param treatment numeric vector of treatment identifiers
#' @param outcome numeric vector of outcomes
#' @param prediction numeric vector of uplift predictions
#' @param p optional "infer", numeric, numeric vector representing treatment
#'   propensities
#' @param n_bins integer number of bins on x-axis; default 20
#' @return a list representing a pylift uplift eval object
#'
#' @examples
#'
#' set.seed(0)
#' rl <- function(x){
#'   round(1/(1+exp(-x)))
#' }
#' n <- 2000; p <- 3
#' beta <- -0.5
#' X <- matrix(rnorm(n*p), n, p)
#' W <- rbinom(n, 1, 0.5)
#' Y <- rl(pmax(beta+X[,1], 0) * W + X[,2])
#' p1 <- 1/(1+exp(-(beta+X[,1])))
#' plUpliftEval(W, Y, p1)
#'
#' \donttest{
#' library(grf)
#' set.seed(123)
#'
#' rl <- function(x){
#'   round(1/(1+exp(-x)))
#' }
#' n <- 2000; p <- 10
#' X <- matrix(rnorm(n*p), n, p)
#' W <- rbinom(n, 1, 0.2)
#' Y <- rl(rl(X[,1]) * W - rl(X[,3]) * W + rnorm(n))
#' tau.forest <- causal_forest(X, Y, W)
#' tau.hat <- predict(tau.forest, X)
#' plue <- plUpliftEval(W, Y, tau.hat$predictions)
#' plue
#' }
#'
#' @export
plUpliftEval <- function(treatment, outcome, prediction, p = "infer", n_bins = 20){

  stopifnot(length(treatment)==length(outcome) & length(treatment)==length(prediction))
  stopifnot(is.numeric(treatment))
  stopifnot(is.numeric(outcome))
  stopifnot(is.numeric(prediction))
  stopifnot(n_bins%%1==0)

  # Deal with `p`, in case float or None.
  if(is.character(p)){
    if(p=="infer"){
      p <- rep(1,length(prediction))*length(treatment[treatment==1])/length(treatment)
    }
  } else if (length(p)==1 & is.numeric(p)){
    p <- rep(1,length(prediction))*p
  } else{
    stopifnot(length(p)==length(prediction) & is.numeric(p))
  }

  new_PlUpliftEval(treatment, outcome, prediction, p)

}


pl_calc <- function(self, plot_type, n_bins=20){

  # Create bins.
  bin_range = seq(0, length(self$treatment), length.out=n_bins+1)

  # Sort `self.prediction`, descending, then get the indices in the test set that
  # these correspond to.
  prob_index = order(self$prediction,decreasing=TRUE)

  # Define whether the curve uses all data up to the percentile, or the data within that percentile.
  noncumulative_subset_func <- function(i){
    seq(1:length(self$treatment)) %in% prob_index[bin_range[i]:bin_range[i+1]]
  }
  cumulative_subset_func <- function(i){
    seq(1:length(self$treatment)) %in% prob_index[1:bin_range[i+1]]
  }

  subsetting_functions = list(
    'qini' = cumulative_subset_func,
    'aqini' = cumulative_subset_func,
    'cgains' = cumulative_subset_func,
    'cuplift' = cumulative_subset_func,
    'balance' = noncumulative_subset_func,
    'uplift' = noncumulative_subset_func
  )

  # Define the function that is calculated within the above bins.
  y_calculating_functions = list (
    'qini' = function(nt1o1, nt0o1, nt1, nt0){nt1o1/self$counts$N_treat - nt0o1/self$counts$N_contr},
    'aqini' = function(nt1o1, nt0o1, nt1, nt0){nt1o1/self$counts$N_treat - nt0o1*nt1/(nt0*self$counts$N_treat + EPS)},
    'cgains' = function(nt1o1, nt0o1, nt1, nt0){(nt1o1/(nt1+EPS)- nt0o1/(nt0+EPS))*(nt1+nt0)/self$counts$N},
    'cuplift' = function(nt1o1, nt0o1, nt1, nt0){nt1o1/(nt1+EPS) - nt0o1/(nt0+EPS)},
    'uplift' = function(nt1o1, nt0o1, nt1, nt0){ nt1o1/(nt1+EPS) - nt0o1/(nt0+EPS)},
    'balance' = function(nt1o1, nt0o1, nt1, nt0){nt1/(nt0+nt1+EPS)}
  )

  # Initialize output lists.
  x = c()
  y = c()

  # Calculate qini curve points for each bin.
  for (i in 1:n_bins){
    current_subset = subsetting_functions[[plot_type]](i)
    # Get the values of outcome in this subset for test and control.
    treated_subset = (self$treatment==1) & current_subset
    resp_treated = self$outcome[treated_subset]
    untreated_subset = (self$treatment==0) & current_subset
    resp_untreated = self$outcome[untreated_subset]
    # Get the policy for each of these as well.
    p_treated = self$p[treated_subset]
    p_untreated = self$p[untreated_subset]

    # Count the number of correct values (i.e. y==1) within each of these
    # sections as a fraction of total ads shown.
    nt1o1 = sum(resp_treated*0.5/p_treated)
    nt0o1 = sum(resp_untreated*0.5/(1-p_untreated))
    nt1 = sum(0.5/p_treated)
    nt0 = sum(0.5/(1-p_untreated))
    y = c(y, y_calculating_functions[[plot_type]](nt1o1, nt0o1, nt1, nt0))
    x = c(x,nt1+nt0)
  }

  # For non-cumulative functions, we need to do a cumulative sum of the x
  # values, because the sums in the loop only captured the counts within
  # the non-cumulative bins.
  if (plot_type %in% c('balance', 'uplift')){
    x = cumsum(x)
  }

  # Rescale x so it's between 0 and 1.
  percentile = x/max(x)

  if (! plot_type %in% c('balance', 'uplift', 'cuplift')){
    percentile = c(0,percentile)
    y = c(0,y)
  }

  list(
    x=percentile,
    y=y
  )
}

#' A port of pylift's plot function (https://github.com/wayfair/pylift) as of commit:
#' https://github.com/wayfair/pylift/tree/bb69692388b1fe085001c3ba7edf6dd81d888353
#
#' pylift: Plots the different kinds of percentage-targeted curves.
#'
#' @param plue the result of a call to the plUpliftEval constructor
#' @param plot_type string, optional Either 'qini', 'aqini', 'uplift', 'cuplift', or 'balance'.
#'     'aqini' refers to an adjusted qini plot, 'cuplift' gives a
#'     cumulative uplift plot. 'balance' gives the test-control balance
#'     for each of the bins. All others are self-explanatory.
#' @param n_bins integer, number of population bins; default 20
#' @param show_theoretical_max boolean, optional
#'     Toggle theoretical maximal qini curve, if overfitting to
#'     treatment/control. Only works for Qini-style curves.
#' @param show_practical_max boolean, optional
#'     Toggle theoretical maximal qini curve, if not overfitting to
#'     treatment/control. Only works for Qini-style curves.
#' @param show_no_dogs boolean, optional
#'     Toggle theoretical maximal qini curve, if you believe there are no
#'     sleeping dogs. Only works for Qini-style curves.
#' @param show_random_selection boolean, optional
#'     Toggle straight line indicating a random ordering. Only works for
#'     Qini-style curves.
#' @param ... additional arguments
#' @return a pylift plot
#'
#' @examples
#'
#' set.seed(0)
#' rl <- function(x){
#'   round(1/(1+exp(-x)))
#' }
#' n <- 2000; p <- 3
#' beta <- -0.5
#' X <- matrix(rnorm(n*p), n, p)
#' W <- rbinom(n, 1, 0.5)
#' Y <- rl(pmax(beta+X[,1], 0) * W + X[,2])
#' p1 <- 1/(1+exp(-(beta+X[,1])))
#' plue <- plUpliftEval(W, Y, p1)
#' pl_plot(plue,
#'         show_practical_max = TRUE,
#'         show_theoretical_max = TRUE,
#'         show_no_dogs = TRUE,
#'         n_bins=20)
#'
#' \donttest{
#' library(grf)
#' set.seed(123)
#'
#' rl <- function(x){
#'   round(1/(1+exp(-x)))
#' }
#' n <- 2000; p <- 10
#' X <- matrix(rnorm(n*p), n, p)
#' W <- rbinom(n, 1, 0.2)
#' Y <- rl(rl(X[,1]) * W - rl(X[,3]) * W + rnorm(n))
#' tau.forest <- causal_forest(X, Y, W)
#' tau.hat <- predict(tau.forest, X)
#' plue <- plUpliftEval(W, Y, tau.hat$predictions)
#' plue
#' pl_plot(plue,
#'         show_practical_max = TRUE,
#'         show_theoretical_max = TRUE,
#'         show_no_dogs = TRUE,
#'         n_bins=20)
#' }
#'
#' @export

pl_plot <- function(plue,
                    plot_type='cgains',
                    n_bins=20,
                    show_theoretical_max=FALSE,
                    show_practical_max=FALSE,
                    show_random_selection=TRUE,
                    show_no_dogs=FALSE,
                    ...){
  print(paste0("plotting: ", plot_type))

  stopifnot(plot_type %in% PLOT_TYPES)

  # Calculate curve (qini, uplift, cumulative uplift).
  xy <- pl_calc(plue, plot_type=plot_type, n_bins=n_bins)

  titles <- list(
    qini = 'Qini curve',
    aqini = 'Adjusted Qini curve',
    cgains = 'Cumulative gain chart',
    cuplift = 'Cumulative uplift curve',
    uplift = 'Uplift curve',
    balance = 'Treatment balance curve'
  )

  ylabels <- list(
    qini = 'Uplift gain',
    aqini = 'Uplift gain',
    cgains = 'Uplift gain',
    cuplift = 'Cumulative lift',
    uplift = 'Lift',
    balance = 'Treatment size / (treatment size + control size)'
  )

  if ((plot_type=='aqini') | (plot_type=='cgains') | (plot_type=='qini')){
    max_plot_type='qini'
  }
  else {
    max_plot_type=plot_type
  }

  plot <- ggplot()+
    geom_line(data=as.data.frame(xy),aes(x,y)) +
    xlab("Fraction of data") +
    ylab(ylabels[[plot_type]]) +
    labs(
      title=titles[[plot_type]]
    ) +
    theme(plot.title = element_text(size=8))

  if(show_random_selection){
    plot <- plot + geom_line(data=data.frame(x=c(0,1),y=c(0,xy$y[length(xy$y)])),aes(x,y), color="grey", linetype="dotted")
  }

  if (plot_type!='balance'){
    if (show_theoretical_max){
      xys <- plue$curve_count_xys[[paste0(max_plot_type,'_max')]]
      plot <- plot +
        geom_line(data=data.frame(x=xys$x,y=xys$y),aes(x,y), color="red", linetype="dotted")
    }
    if (show_practical_max){
      xys <- plue$curve_count_xys[[paste0(max_plot_type,'_pmax')]]
      plot <- plot +
        geom_line(data=data.frame(x=xys$x,y=xys$y),aes(x,y), color="blue", linetype="dotted")
    }
    if (show_no_dogs){
      xys <- plue$curve_count_xys[[paste0(max_plot_type,'_nosdmax')]]
      plot <- plot +
        geom_line(data=data.frame(x=xys$x,y=xys$y),aes(x,y), color="green", linetype="dotted")
    }
  }

  plot

}
